File: packages/capacitor/android/src/main/java/com/nathanclaire/rum/EdgeRumCrashPlugin.kt
package com.nathanclaire.rum

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * Capacitor plugin shim. Thin — delegates to:
 *  - [JvmCrashHandler] for Java/Kotlin uncaught Throwables
 *  - [AnrWatchdog] for main-thread liveness
 *  - [NativeCrashBridge] for NDK signal handlers
 *
 * NOTE: this file has not been built in this repo's CI — see
 * `goofy-rolling-flask.md` "Manual validation steps" for the smoke tests
 * you need to run on a real Android device or emulator.
 */
@CapacitorPlugin(name = "EdgeRumCrash")
class EdgeRumCrashPlugin : Plugin() {

    @Volatile private var installed = false
    private val storage by lazy { CrashRecordStorage(context.cacheDir) }

    @PluginMethod
    fun install(call: PluginCall) {
        if (installed) {
            call.resolve(JSObject().put("installed", true))
            return
        }

        val enableAnr = call.getBoolean("enableAnrDetection", true) ?: true
        val anrTimeoutMs = call.getInt("anrTimeoutMs", 5000) ?: 5000

        try {
            JvmCrashHandler.install(storage) { lastScreen() }
            NativeCrashBridge.install(storage.ndkRecordFile.absolutePath) { lastScreen() }
            if (enableAnr) {
                AnrWatchdog.start(storage, anrTimeoutMs) { lastScreen() }
            }
            installed = true
            call.resolve(JSObject().put("installed", true))
        } catch (t: Throwable) {
            call.resolve(JSObject().put("installed", false).put("error", t.message ?: ""))
        }
    }

    @PluginMethod
    fun fetchPending(call: PluginCall) {
        val jvm = storage.readJvmCrashes()
        val anr = storage.readAnrCrashes()
        val ndk = NativeCrashBridge.fetchPendingNdk(storage)
        val all = (jvm + anr + ndk)
        val arr = JSArray()
        for (r in all) arr.put(r.toJson())
        call.resolve(JSObject().put("crashes", arr))
    }

    @PluginMethod
    fun markHandled(call: PluginCall) {
        val ids = call.getArray("ids")
        val idList = mutableListOf<String>()
        if (ids != null) {
            for (i in 0 until ids.length()) {
                val s = ids.optString(i, null) ?: continue
                idList.add(s)
            }
        }
        storage.deleteByIds(idList)
        call.resolve()
    }

    @PluginMethod
    fun setLastScreen(call: PluginCall) {
        val screen = call.getString("screen", "") ?: ""
        storage.setLastScreen(screen)
        call.resolve()
    }

    @PluginMethod
    fun startPerfSampling(call: PluginCall) {
        val captureFrames = call.getBoolean("captureFrames", true) ?: true
        val captureMemory = call.getBoolean("captureMemory", true) ?: true
        val memoryIntervalMs = call.getInt("memoryIntervalMs", 10_000) ?: 10_000
        // `getDouble` is missing on PluginCall; JS sends a number, JSON parses
        // as Double, getString round-trip is the simplest cross-bridge option.
        val slowThresholdMs = call.data.optDouble("frameSlowThresholdMs", 16.67)
        val captureAllFrames = call.getBoolean("captureAllFrames", false) ?: false

        if (captureMemory) {
            MemorySampler.start(context, memoryIntervalMs.toLong())
        }
        if (captureFrames) {
            FrameSampler.start(slowThresholdMs, captureAllFrames) { lastScreen() }
        }
        call.resolve(JSObject().put("started", true))
    }

    @PluginMethod
    fun stopPerfSampling(call: PluginCall) {
        runCatching { MemorySampler.stop() }
        runCatching { FrameSampler.stop() }
        call.resolve()
    }

    // Backgrounding is a scene change: close the current frame window so the
    // background gap isn't measured as one giant dropped frame on resume
    // (mirrors iOS didEnterBackground). No-op unless frame sampling is running.
    override fun handleOnPause() {
        super.handleOnPause()
        runCatching { FrameSampler.onBackground() }
    }

    @PluginMethod
    fun fetchFrameSamples(call: PluginCall) {
        val frames = runCatching { FrameSampler.fetchPending() }.getOrDefault(emptyList())
        val arr = JSArray()
        for (f in frames) arr.put(toJsObject(f))
        call.resolve(JSObject().put("frames", arr))
    }

    @PluginMethod
    fun fetchMemorySamples(call: PluginCall) {
        val samples = runCatching { MemorySampler.fetchPending() }.getOrDefault(emptyList())
        val arr = JSArray()
        for (s in samples) arr.put(toJsObject(s))
        call.resolve(JSObject().put("samples", arr))
    }

    private fun toJsObject(map: Map<String, Any?>): JSObject {
        val obj = JSObject()
        for ((k, v) in map) {
            // JSObject has overloaded put() for primitives; nulls are dropped
            // (matches the wire contract — omit instead of sending null).
            when (v) {
                null -> Unit
                is Boolean -> obj.put(k, v)
                is Int -> obj.put(k, v)
                is Long -> obj.put(k, v)
                is Double -> obj.put(k, v)
                is Float -> obj.put(k, v.toDouble())
                else -> obj.put(k, v.toString())
            }
        }
        return obj
    }

    private fun lastScreen(): String = storage.getLastScreen()
}

File: packages/capacitor/ios/Plugin/EdgeRumCrashPlugin.swift
import Foundation
import Capacitor

/// Capacitor plugin bridge. Thin shim — delegates to `CrashReporter` (PLCrashReporter wrapper)
/// and `HangDetector` for the actual work.
///
/// NOTE: this file ships in `packages/capacitor/ios/` and is built via CocoaPods when the
/// consumer app runs `pod install`. It has not been built in this repo's CI — see
/// `goofy-rolling-flask.md` "Manual validation steps" for the smoke tests you need to run.
@objc(EdgeRumCrashPlugin)
public class EdgeRumCrashPlugin: CAPPlugin {

    private static var installed = false
    private static let installLock = NSLock()

    /// Called once by Capacitor when the plugin is discovered and linked,
    /// before any JS runs. Set a breakpoint here (or grep the device console
    /// for the `[edge-rum]` debug log) to verify that SPM/CocoaPods linkage
    /// is wiring this plugin into the consumer app. The real work happens
    /// in `install(_:)` which fires when JS calls `plugin.install()`.
    public override func load() {
        super.load()
        #if DEBUG
        print("[edge-rum] EdgeRumCrashPlugin loaded (SPM/CocoaPods linkage OK; awaiting JS install)")
        #endif
    }

    @objc func install(_ call: CAPPluginCall) {
        EdgeRumCrashPlugin.installLock.lock()
        defer { EdgeRumCrashPlugin.installLock.unlock() }

        if EdgeRumCrashPlugin.installed {
            call.resolve(["installed": true])
            return
        }

        let enableHang = call.getBool("enableHangDetection") ?? true
        let hangTimeoutMs = call.getInt("hangTimeoutMs") ?? 5000

        do {
            try CrashReporter.shared.install()
        } catch {
            // PLCrashReporter failed to arm — still mark as installed=false so the JS
            // bridge surfaces a health-monitor error but doesn't loop on retry.
            call.resolve(["installed": false, "error": error.localizedDescription])
            return
        }

        if enableHang {
            HangDetector.shared.start(timeoutMs: hangTimeoutMs)
        }

        EdgeRumCrashPlugin.installed = true
        call.resolve(["installed": true])
    }

    @objc func fetchPending(_ call: CAPPluginCall) {
        let crashes = CrashReporter.shared.fetchPending()
        call.resolve(["crashes": crashes])
    }

    @objc func markHandled(_ call: CAPPluginCall) {
        guard let ids = call.getArray("ids", String.self) else {
            call.reject("`ids` must be an array of strings")
            return
        }
        CrashReporter.shared.markHandled(ids: ids)
        call.resolve()
    }

    @objc func setLastScreen(_ call: CAPPluginCall) {
        let screen = call.getString("screen") ?? ""
        CrashReporter.shared.setLastScreen(screen)
        call.resolve()
    }
}
